Enable Kerberos auth for DCERPC

smashery opened the pull request:
This PR supports the kerberos-in-DCSync work in the Metasploit repo (see: https://github.com/rapid7/metasploit-framework/pull/18419). Test cases are listed in that PR, and should exercise all of the changes in this PR.

smcintyre-r7 commented on lib/ruby_smb/dcerpc/response.rb:
```diff
@@ -46,7 +44,7 @@ def stub_length
       end
 
-      def read(io)
+       def read(io)
```
Looks like a stray space was added here.

smcintyre-r7 commented on lib/ruby_smb/dcerpc/response.rb:
```diff
@@ -16,9 +16,7 @@ class Response < BinData::Record
       string      :stub, label: 'Stub', read_length: -> { stub_length }
-      string      :auth_pad,
-        onlyif: -> { has_auth_verifier? },
-        length: -> { (16 - (stub.num_bytes % 16)) % 16 }
+      string      :auth_pad, onlyif: -> { has_auth_verifier? }
```
This change here and in the request is a bit concerning because it'd alter the behavior. If I understand correctly though from my review, it's necessary because the padding now needs to be calculated by the Kerberos encryptor when GSS authentication is in use. Is that right?

If that's the case, it still seems like there should be a read_length defined for the string field here.

smashery replied:
So my understanding is that the `read_length` attribute is used in the situation of reading into the structure with the `read` method - i.e. telling it how many bytes to read for that field. In the `Response` class, however, the `read` function has been explicitly overridden to manually fill the `auth_pad` field based on the `sec_trailer.auth_pad_length` field.

Regarding the rationale for removing the `length` field, this was only used (as far as I can tell) in automatically setting the `auth_pad_length` field in the `sec_trailer`; and *that* is the place where behaviour has definitely been changed.

From a high level, it seemed odd to me that the "length" of a bytestring was based not on its actual length, but on the length of a different bytestring. That somehow seems backwards. Certainly it would make sense to have a convenience function (i.e. `how_long_should_the_auth_pad_be`), but automatically setting the "length" of a string based on the length of another string feels non-obvious, as a reader of the code. It also allows the situation where `auth_pad_length` is saying one thing, but the caller can set `auth_pad`'s actual length to something different; so the existing solution isn't a panacea.

In practice, this "magic" made it much harder to reason through the code while debugging a separate issue I was having, and, while it may be strictly correct, it caused all sorts of confounding factors that made the code hard to debug. Once I realised this magic was happening, I was able to make some progress; but as soon as I removed this automated behaviour, and took responsibility as the caller for explicitly calculating and setting the value, I found it so much clearer to read the code, and much easier to reason about what was happening, rather than having values unexpectedly setting themselves implicitly behind my back, and having to add that to my mental model of the code.

This admittedly does put the onus on the caller to ensure the struct has valid padding. But as mentioned earlier, there was already some onus on the caller (to ensure that the `auth_pad` value's actual length matched the automatically-calculated `sec_trailer.auth_pad_length` value); so it's really a trade-off of "code is clearer, but the caller needs to set two values correctly, rather than just one".

To help ensure correctness, I added an assertion in `client.rb`, which will throw an error if the multiple-of-16 requirement is violated. I've just added an extra assertion that `auth_pad_length` is equal to `auth_pad.length` - will push that up in a sec.

So the tl;dr here is: I found the code way more readable once I removed it, it never guaranteed correctness anyway, and we're now asserting on malformed structures, so should catch incorrect uses.

But if we're worried about this being a breaking change for other users of `ruby_smb`, I can revert the change to the automated approach.

smashery replied:
Okay, I've restored it to the old behaviour, and made that change to make it a bit more readable.

cdelafuente-r7 reviewed (commented):
Thanks @smashery for updating the library! This is a great thing to add support to Kerberos authentication in MSF, awesome work! My only regret is that it is not implemented in RubySMB itself. That would have been even better and many RubySMB users would have benefit from this. But I understand the limitations and all the Kerberos dependencies to MSF.

That being said, I left a few comments for you to review when you get a chance. One main issue I saw during testing was the signature validation. Please, let me know if you have any questions.

smashery commented on lib/ruby_smb/dcerpc.rb:
```diff
@@ -74,23 +174,16 @@ def bind(options={})
-          raise ArgumentError, "Unsupported Auth Type: #{options[:auth_type]}"
-        end
-        add_auth_verifier(bind_req, auth, options[:auth_type], options[:auth_level])
+      force_set_auth_params(auth_type, auth_level)
```
This is mostly for the unit tests; because I changed the interface to have the auth type and level set on the class itself, rather than in each method. The result is that it's more difficult to use the class incorrectly (e.g. changing auth type throughout the exchange, which is just not meaningful); but it does mean that test setup would require calling `bind`. So I added this to skip the `bind` setup in tests.

smashery commented on lib/ruby_smb/dcerpc.rb:
```diff
@@ -56,6 +56,106 @@ module Dcerpc
+    # This function should be overriden for other providers (e.g. Kerberos, etc.)
+    #
+    # @param response [BindAck] the BindAck response packet
+    def auth_provider_complete_handshake(response, options)
```
Options is needed for when Kerberos overrides it. I've documented that in a `@param` comment.

smashery commented on lib/ruby_smb/dcerpc.rb:
```diff
@@ -56,6 +56,106 @@ module Dcerpc
+      if @ntlm_client.flags & NTLM::NEGOTIATE_FLAGS[:EXTENDED_SECURITY] != 0
+        data_to_check = dcerpc_response.to_binary_s[0..-(dcerpc_response.pdu_header.auth_length + 1)]
+      end
+      valid = @ntlm_client.session.verify_signature(signature, data_to_check)
```
Ah, of course - I've taken it outside of the `if` block.

smashery commented on lib/ruby_smb/dcerpc.rb:
```diff
@@ -214,39 +307,19 @@ def set_integrity_privacy(dcerpc_req, auth_level:, auth_type:)
-        data_to_check = dcerpc_response.to_binary_s[0..-(dcerpc_response.pdu_header.auth_length + 1)]
-      end
-      unless @ntlm_client.session.verify_signature(signature, data_to_check)
+      unless signature_valid
```
Thanks - it was missing including the auth pad in the verification - I've amended this.

smashery commented:
Thanks for the review @cdelefuente-r7. Agreed regarding having the Kerberos functionality available here. I wonder whether it would be worth a refactoring job to pull it out into either here, or a separate repo?
I think I've addressed the comments - let me know if there's anything else needed.

cdelafuente-r7 reviewed (commented):
Sorry for the last minute comment before it lands. I just noticed a few comments from last week were not answered yet. Also, I added one more about the padding length calculation logic in the request.

smcintyre-r7 commented:
@cdelafuente-r7 I opened a PR implemented the changes you requested here https://github.com/smashery/ruby_smb/pull/1. Give it a look and let me know what you think

smcintyre-r7 commented:
Tested this with the Metasploit side of things and it's all looking good to me now!

Thanks for your work on this Smashery, this is a great improvement!

```
msf6 auxiliary(gather/windows_secrets_dump) > run DOMAIN=msflab.local SMB::Auth=kerberos SMB::Rhostname=dc.msflab.local DomainControllerRhost=192.168.159.10
[*] Running module against 192.168.159.10

[+] 192.168.159.10:445 - 192.168.159.10:88 - Received a valid TGT-Response
[*] 192.168.159.10:445 - 192.168.159.10:445 - TGT MIT Credential Cache ticket saved to /home/smcintyre/.msf4/loot/20231025135757_default_192.168.159.10_mit.kerberos.cca_191490.bin
[+] 192.168.159.10:445 - 192.168.159.10:88 - Received a valid TGS-Response
[*] 192.168.159.10:445 - 192.168.159.10:445 - TGS MIT Credential Cache ticket saved to /home/smcintyre/.msf4/loot/20231025135757_default_192.168.159.10_mit.kerberos.cca_964676.bin
[+] 192.168.159.10:445 - 192.168.159.10:88 - Received a valid delegation TGS-Response
[*] 192.168.159.10:445 - Service RemoteRegistry is already running
[*] 192.168.159.10:445 - Retrieving target system bootKey
[+] 192.168.159.10:445 - bootKey: 0x369c37bc5ec5e4b3eaeee7b69caf5d6a
[*] 192.168.159.10:445 - Saving remote SAM database
[*] 192.168.159.10:445 - Dumping SAM hashes
[*] 192.168.159.10:445 - Password hints:
No users with password hints on this system
[*] 192.168.159.10:445 - Password hashes (pwdump format - uid:rid:lmhash:nthash:::):
... cut because secrets
```
